MDANSE/Src/MDANSE/Chemistry/GroupingTool.py

```diff
@@ -0,0 +1,415 @@
+            information about atoms and molecules in the system
+        output_data : OutputData
+            structure into which to write the results of the calculation
+
+        """
+        self._cs = chemical_system
+        self._output_data = output_data
+        self._grouping = None
+        self._data_parameters = {}
+        self._weight_dictionary = {}
+        self._mass_dictionary = {}
+        self._molecule_mass = defaultdict(float)
+        self._current_selection = set()
+        self._indices_per_data_key = {}
+        self._mandatory_keys = ["output_type", "dimensions"]
+        self._extra_keys = ["axis", "units", "main_result", "partial_result", "dtype"]
+        self._plain_datasets = set()
+        self._weighted_datasets = set()
+        self._molecule_datasets = set()
+        self._mutex = Lock()
+        self._debug_counter = Counter()
+
+    def set_grouping(self, text_key: str):
+        """Choose the level of grouping to use.
+
+        This has to be configured before creating output data sets.
+
+        Parameters
+        ----------
+        text_key : str
+            name of the grouping level in GROUPING_LABELS
+
+        """
+        self._grouping = GROUPING_LABELS.get(text_key, GroupingLevel.ATOM)
+
+    def set_weight_dictionary(self, new_weights: dict[tuple[str], float]):
+        """Store the dictionary with weights per atom type.
+
+        Parameters
+        ----------
+        new_weights : dict[tuple[str],float]
+            dictionary of {atom_types: scaling_factor} pairs
+
+        """
+        self._weight_dictionary = {
+            "".join([*key]): value for key, value in new_weights.items()
+        }
+
+    def set_atom_masses(self, atom_database):
+        """Get atom masses from an object that has .get_atom_property.
+
+        Parameters
+        ----------
+        atom_database : Union[AtomDatabase, Trajectory]
+            any object that has get_atom_property
+
+        """
+        self._mass_dictionary = {
+            atom_type: atom_database.get_atom_property(atom_type, "atomic_weight")
+            for atom_type in self._cs._unique_elements
+        }
+
+    def set_dataset_parameters(self, parameters: dict[str, Any]):
+        """Save information about axes, units and array dimensions.
+
+        These will be used for all the output data.
+
+        Parameters
+        ----------
+        parameters : dict[str, Any]
+            dictionary of parameters needed by OutputData.write
+
+        """
+        for key in self._mandatory_keys:
+            if key not in parameters:
+                raise KeyError(f"Setting {key} is missing from data parameters.")
+        for key in self._mandatory_keys:
+            self._data_parameters[key] = parameters[key]
+        for key in self._extra_keys:
+            value = parameters.get(key)
+            if value is None:
+                value = BACKUP_DATA_PARAMETERS[key]
+            self._data_parameters[key] = value
```

Comment:
In fact there's no need to have two loops here. You've guaranteed the mandatory keys are in the keys, so we can instead just make this:

```suggestion
        for key in self._mandatory_keys + self._extra_keys:
            self._data_parameters[key] = parameters.get(key, BACKUP_DATA_PARAMETERS.get(key))
```